### DISCUSSION TITLE
PCF2022 with MDI TabbedView seems to have a few resize bugs

### DISCUSSION THREAD

**Discussion by BenchTop2007** [OpenSourcePFCLibraries/2022#6] [Q&A] [answered]
Possible pfc u_lv bug in MDI TabbedView 
MDI application starts in a maximized state.
User successfully logs in.
The Navigator Window opens automatically. Tree View on left, List View on right
Navigator list view's View property is set to listviewlargeicon! 
Open event registers the tree view and list view for the resize service (inv_resize.of_Register)

Next, click the button next to the X on the top, right side of the MDI frame to set the window size to "normal" (not maximized)
In the Navigator's List View, the images don't re-arrange and horizontal/vertical scroll bars do not display.
Grab the bottom right side of the MDI frame and resize the frame slightly.
Now as I select different items in the tree view, the list view icons all re-arrange and display properly.
I must manually resize the frame slightly to get the list view icons to re-arrange to fit the window.

Similar behavior with u_tab objects as described above.
When the MDI frame is not maximized, open another window from the Navigator.
Maximize the MDI frame
The u_dw and u_tab objects do not resize.
Set the MDI frame back to normal window size (not maximized)
Grab the bottom right side of the MDI frame and resize the frame slightly
Vertical/Horizontal scoll bars display correctly.
Maximize the MDI frame again
The u_dw and u_tab objects resized correctly.

Using PB2022 R2.

**Comment by BenchTop2007** [comment]
I create a small project to demonstrate what's going on... attached zip file. 
[pfcbugs.zip](https://github.com/OpenSourcePFCLibraries/2022/files/13548985/pfcbugs.zip)

To replicate:

1. Run the project.  The MDI frame opens in a maximized state.

2. Click the box next to X on the top, right side of the frame to put the frame in a normal state, not maximized.

3. Select File - Open.  A window with a datawindow will display. Horizontal and vertical scroll bars display.

4. Click the box next to X on the top, right side of the frame to put the frame back in a maximized state.

5. The datawindow does not resize. 

6. When I put the frame back into a normal state, then maximized state again, the datawindow does resize.

**Comment by bruce-armstrong** [comment]
The datawindow does not resize because when you maximized the frame no resize notifications were sent to the sheets.  There are a number of other issues that make it difficult at this time to try to address this and some other resize issues.  I've sent this note to Appeon in the hopes that they can address some of these and make it more practical for us to address or eliminate the need for us to address it at all.

1.	When tabbedview MDI is used and the tabs are maximized, when the frame size is changed the sheets do not resize.  I believe given that they are maximized to fit the frame that they should be automatically resized when the frame is resized.
2.	When tabbedview MDI is used and the tabs are not maximized, a resize even is *not* sent to the window when a sheet is maximized.  It is when it’s resized manually or minimized, but not when it’s maximized.  I believe this is a bug.
3.	When tabbedview MDI is used and the tabs are maximized, the resize event and this.WorkSpaceHeight() both return a value that represents the entire MDI_1 client area.  However, the tabbed view control (mditbb_1) is occupying bar of that space.  In order to determine what is the actual useful window area, the height of mdittb_1 has to be subtracted from the newsize/WorkSpaceHeight when the sheet is maximized.
4.	In order to try to adjust the resize behavior of PFC I need to be able to determine dynamically whether tabbedview MDI is in use and whether it has been configured to maximize the sheets.  However:
a.	Not only are the TabbedView and MaximizeAllTabbedSheets not setable at runtime, they aren’t readable either.  There are no properties reported on the window type to correspond to them.
b.	I could try to determine if TabbedView is in use by trying to determine if mditbb_1 is valid at runtime.  However that doesn’t appear as a property I can access from outside of the frame (it’s not a property on the window class).  Note that mdi_1 is a property of the window class, so that technique does work for mdi_1, but not for mditbb_1.
c.	What I’ve found does work is looping through the control array of the frame and looking to see if one of the objects has a classname of mditbb_1.  However that only tells me that TabbedMDI is in use, there still doesn’t seem to be any way that I can determine whether or not MaximizeAllTabbesSheets is set.

**Reply by BenchTop2007** [reply]
Thank you Bruce!

**Comment by BenchTop2007** [comment]
Bruce,

I see PB2022 R3 GL came out yesterday... any idea how long it will take to update the PFC?  No rush, just wondering if it will be weeks or months?

~~~Tracy

**Comment by bruce-armstrong** [comment, marked as answer]
No time at all.  I tested this during the R3 beta and, at least with the test case I had, no changes were necessary.  Once PB started sending the missing resize notifications the PFC resize service handled the resizes correctly.
